components/context_id/src/error.rs

```diff
@@ -0,0 +1,40 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use error_support::{ErrorHandling, GetErrorHandling};
+
+pub type Result<T> = std::result::Result<T, Error>;
+pub type ApiResult<T> = std::result::Result<T, ApiError>;
+
+#[derive(Debug, thiserror::Error, uniffi::Error)]
+pub enum ApiError {
+    #[error("Something unexpected occurred.")]
+    Other { reason: String },
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Timestamp was invalid")]
+    InvalidTimestamp { timestamp: i64 },
+
+    #[error("URL parse error: {0}")]
+    UrlParseError(#[from] url::ParseError),
+
+    #[error("Viaduct error: {0}")]
+    ViaductError(#[from] viaduct::Error),
+
+    #[error("UniFFI callback error: {0}")]
+    UniFFICallbackError(#[from] uniffi::UnexpectedUniFFICallbackError),
```

Comment:
This looks good to me, although it's a bit weird to think through the flow.  I feel like we could have a better story here, maybe callbacks like yours could return `Result<(), uniffi::UnexpectedUniFFICallbackError`>` and not have to do any of this conversion.

In any case, this is just me thinking out loud how this could be improved from the UniFFI side of things.  I don't think you need to change any of this code.

Reply:
Amazing, thanks!